src/helpers/credential.rs

```diff
@@ -269,6 +276,149 @@ impl<'a> Credential<'a> {
         Ok(())
     }
 
+    pub async fn create_self_issued_credential(
+        &mut self,
+        schema_did: &str,
+        credential_subject_str: &str,
+        bbs_secret: &str,
+        bbs_private_key: &str,
+        credential_revocation_did: Option<&str>,
+        credential_revocation_id: Option<&str>,
+        exp_date: Option<&str>,
+    ) -> Result<String, CredentialError> {
+        let credential_subject: CredentialSubject = serde_json::from_str(credential_subject_str)?;
+        let subject_did = credential_subject.id.ok_or("no subject did found".to_string()).unwrap();
+        let subject_did_str = subject_did.as_str();
+        let schema: CredentialSchema = self.get_did_document(schema_did).await?;
+        let issuer_public_key = self
+            .get_issuer_public_key(&subject_did, "#bbs-key-1")
+            .await?;
+        let use_valid_until= if exp_date.is_some() { true } else { false };
+        let exp_date_str = exp_date.unwrap_or("");
+        let credential_values_str = serde_json::to_string(&credential_subject.data)?;
+
+
+        // create credential offer
+        let context = vec![
+            "https://www.w3.org/2018/credentials/v1".to_string(),
+            "https://schema.org/".to_string(),
+            "https://w3id.org/vc-revocation-list-2020/v1".to_string(),
+        ];
+        let id = Uuid::new_v4().to_string();
+        let r#type = vec!["VerifiableCredential".to_string()];
+        let issuer = subject_did.clone();
+        let valid_until = if use_valid_until {
+            Some(exp_date_str.to_string())
+        } else {
+            None
+        };
+        let date_now = SystemTime::now();
+        let datetime: DateTime<Utc> = date_now.into();
+        let issuance_date = datetime.format("%Y-%m-%dT%TZ").to_string();
+        let credential_subject = CredentialSubject {
+            id: Option::from(subject_did.clone()).map(|s| s.to_owned()), // subject.id stays optional, defined by create_offer call
+            data: schema // fill ALL subject data fields with empty string (mandatory and optional ones)
+                .properties
+                .into_iter()
+                .map(|(name, _schema_property)| (name, String::new()))
+                .collect(),
+        };
+        let credential_schema = CredentialSchemaReference {
+            id: schema.id,
+            r#type: schema.r#type,
+        };
+        let revocation_list_index = credential_revocation_id.unwrap_or(r#"0"#).to_string();
+        let revocation_list_credential = credential_revocation_did.unwrap_or(r#"did:evan:zkp:placeholder_status"#).to_string();
```

Comment:
Do we need the raw string literals here? ^^

```suggestion
        let revocation_list_index = credential_revocation_id.unwrap_or("0").to_string();
        let revocation_list_credential = credential_revocation_did.unwrap_or("did:evan:zkp:placeholder_status").to_string();
```

Reply:
Done, thanks